cmd/zfs/zfs_main.c

```diff
@@ -727,6 +727,38 @@ finish_progress(char *done)
 	pt_header = NULL;
 }
 
+/* This function checks if the passed fd refers to /dev/null or /dev/zero */
+#ifdef __linux__
+static int
+is_dev_files(int fd)
+{
+	int nfd = open("/dev/null", O_WRONLY);
+	struct stat st, nst;
+	fstat(fd, &st);
+	fstat(nfd, &nst);
+	int res = (st.st_dev == nst.st_dev);
+	close(nfd);
+	if (res != 1) {
+		nfd = open("/dev/zero", O_WRONLY);
+		fstat(nfd, &nst);
+		res = (st.st_dev == nst.st_dev);
```

Comment:
I don't think this is right: you're comparing st_dev, which is the containing filesystem. You can trivially make `/dev/zero` somewhere else. You probably want to compare `st_rdev`, which is the actual device number. Indeed, you likely want to just do:
```
fstat(fd, &st);
return (major(st.st_rdev) == 1 && (minor(st.st_rdev) == 3 /* null */ || minor(st.st_rdev) == 5 /* zero */);
```

Reply:
That was deliberate; as outlined, the logic implemented was "did we get this error? is it a file in /dev?"

But you're correct, you can go mknod wherever you like. Done.